Content.Shared/Ninja/Systems/EnergyKatanaSystem.cs

```diff
@@ -0,0 +1,100 @@
+using Content.Shared.Charges.Components;
+using Content.Shared.Charges.Systems;
+using Content.Shared.Examine;
+using Content.Shared.Hands.EntitySystems;
+using Content.Shared.Interaction;
+using Content.Shared.Inventory.Events;
+using Content.Shared.Ninja.Components;
+using Content.Shared.Physics;
+using Content.Shared.Popups;
+using Robust.Shared.Audio;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Timing;
+
+namespace Content.Shared.Ninja.Systems;
+
+/// <summary>
+/// System for katana dashing and binding
+/// </summary>
+public sealed class EnergyKatanaSystem : EntitySystem
+{
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly SharedChargesSystem _charges = default!;
+    [Dependency] private readonly SharedHandsSystem _hands = default!;
+    [Dependency] private readonly SharedInteractionSystem _interaction = default!;
+    [Dependency] private readonly SharedSpaceNinjaSystem _ninja = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<EnergyKatanaComponent, GotEquippedEvent>(OnEquipped);
+        SubscribeLocalEvent<NinjaSuitComponent, KatanaDashEvent>(OnDash);
+    }
+
+    /// <summary>
+    /// When equipped by a ninja, try to bind it.
+    /// </summary>
+    private void OnEquipped(EntityUid uid, EnergyKatanaComponent comp, GotEquippedEvent args)
+    {
+        // check if already bound
+        if (comp.Ninja != null)
+            return;
+
+        // check if ninja already has a katana bound
+        var user = args.Equipee;
+        if (!TryComp<SpaceNinjaComponent>(user, out var ninja) || ninja.Katana != null)
+            return;
+
+        // bind it
+        comp.Ninja = user;
+        _ninja.BindKatana(ninja, uid);
+    }
+
+    /// <summary>
+    /// Handle charges and teleport to a visible location.
+    /// </summary>
+    public void OnDash(EntityUid suit, NinjaSuitComponent comp, KatanaDashEvent args)
+    {
+        if (!_timing.IsFirstTimePredicted)
+            return;
+
+        var user = args.Performer;
+        args.Handled = true;
+        if (!TryComp<SpaceNinjaComponent>(user, out var ninja) || ninja.Katana == null)
+            return;
+
+        var uid = ninja.Katana.Value;
+        if (!TryComp<EnergyKatanaComponent>(uid, out var katana) || !_hands.IsHolding(user, uid, out var _))
```

Comment:
Couldn't this just be a DashProviderComponent instead with a user whitelist?

Reply:
yes